### PR TITLE
✨Add features to tool

### PR DESCRIPTION
**Pull request by migbro**
## Description

A popular feature request has been to output old to new gene symbol changes. Also added sort of a hack to define fake column names if the file is header-less.
I renamed the update tool since it's really more generic than it's description and has been used in other context.

Fixes # https://github.com/d3b-center/bixu-tracker/issues/1946

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Local script used to update PVS1.level file
- [ ] Test B

**Test Configuration**:
* Environment:
* Test files:

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [x] I have committed any related changes to the PR

### CODE REVIEW

File: scripts/update_gene_symbols.py [hunk 1]
```diff
@@ -47,16 +47,22 @@ def get_args():
...
             nargs='+',
             default=["FusionName","Gene1A","Gene1B"],
             help="Space-separated column names from the Fusions TSV where to update gene names \
                     (e.g. -u foo bar blah). Default: %(default)s")
+    optional.add_argument(
+            "-z",
+            "--fake_columns",
+            nargs='+',
+            help="Space-separated column names ro create a faux-header if the file is headerless \
+                    (e.g. -u foo bar blah).")
```
Comment (dmiller15) [inline scripts/update_gene_symbols.py]:
```suggestion
            help="Space-separated column names to use as header for Input TSV. Overrides script header detection (e.g. -z foo bar blah).")
```

File: scripts/update_gene_symbols.py [hunk 2]
```diff
@@ -109,27 +115,32 @@ def hgnc_tsv_to_dict(hgnc_file, old_sym, new_sym):
...
+    print ('{}\t{}'.format("old gene", "new gene"), file=sys.stderr)
+    with (gzip.open if input_tsv.endswith("gz") else open)(input_tsv, "rt", encoding="utf-8") as f:
         with (gzip.open if out_file.endswith("gz") else open)(out_file, "wt", encoding="utf-8") as w:
-            header = f.readline().strip().split('\t')
-            w.write('\t'.join(header) + '\n')
+            if fake_columns is None:
+                header = f.readline().strip().split('\t')
+                w.write('\t'.join(header) + '\n')
+            else:
+                header = '\t'.join(fake_columns)
```
Comment (dmiller15) [inline scripts/update_gene_symbols.py:142]: We want to print the provided header to the output, no?
```suggestion
            if fake_columns is None:
                header = f.readline().strip().split('\t')
            else:
                header = '\t'.join(fake_columns)
            w.write('\t'.join(header) + '\n')
```

File: scripts/update_gene_symbols.py [hunk 3]
```diff
@@ -109,27 +115,32 @@ def hgnc_tsv_to_dict(hgnc_file, old_sym, new_sym):
...
 
     Args:
-        fusions_file: An unopened TXT or TXT.GZ file that contains Custom Fusion information
+        input_tsv: An unopened TXT or TXT.GZ file that contains Custom Fusion information
         sym_dict: A dict with old symbol keys and corresponding value list of new symbols
-        update_columns: A list of columnnames in the fusions_file that must be updated
+        update_columns: A list of columnnames in the input_tsv that must be updated
         out_file: A string filename (files ending in GZ will be compressed) for the output
         retain_records: A boolean that indicates if original records should be kept
         explode_records: A boolean that enables multirecord outputs
```
Comment (dmiller15) [inline scripts/update_gene_symbols.py]:
```suggestion
        explode_records: A boolean that enables multirecord outputs
        fake_columns: A list of strings representing the header fields
```

File: tools/update_gene_symbols.cwl [hunk 4]
```diff
@@ -32,11 +32,13 @@ inputs:
     inputBinding: { position: 1, prefix: "--new_symbol"} }
   update_columns: { type: 'string[]?', doc: "Space-separated column names from the Fusions TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults",
     inputBinding: { position: 1, prefix: "--update_columns", shellQuote: false} }
+  fake_columns: { type: 'string[]?', doc: "Space-separated column names ro create a faux-header if the file is headerless (e.g. -u foo bar blah)",
```
Comment (dmiller15) [inline tools/update_gene_symbols.cwl]:
```suggestion
  fake_columns: { type: 'string[]?', doc: "Column names to use as the header for the Input TSV. Set to override script using first line of Input TSV as header or if file has no header (e.g. -z foo bar blah).",
```

File: README.md [hunk 5]
```diff
@@ -17,7 +17,7 @@ In order to make the outputs harmonized with our fusion results we:
  - Update gene symbols
    - `hgnc_tsv`: Gene name [database TSV file](https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt) from HGNC.
   If omitted, this step will be skipped
-   - `fusions_tsv`: Custom Fusions TSV file, i.e. fusion-dgd.tsv.gz }
+   - `input_tsv`: Custom Fusions TSV file, i.e. fusion-dgd.tsv.gz }
    - `old_symbol`: Column name for the old gene symbol(s) in the HGNC TSV. Set to override script defaults
    - `new_symbol`: Column name for the new gene symbol(s) in the HGNC TSV. Set to override script defaults
    - `update_columns`: Column names from the Fusions TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults: `["FusionName","Gene1A","Gene1B"]`
```
Comment (dmiller15) [inline README.md]:
```suggestion
   - `update_columns`: Column names from the Input TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults: `["FusionName","Gene1A","Gene1B"]`
   - `fake_columns`: Column names to use as the header for the Input TSV. Set to override script using first line of Input TSV as header or if file has no header (e.g. -z foo bar blah). 
```

File: scripts/update_gene_symbols.py [hunk 6]
```diff
@@ -47,16 +47,22 @@ def get_args():
...
+            "--input_tsv",
             required=True,
-            help="Custom fusions TSV file: fusion-dgd.tsv.gz")
+            help="Input TSV file: fusion-dgd.tsv.gz")
     optional.add_argument(
             "-u",
             "--update_columns",
             nargs='+',
             default=["FusionName","Gene1A","Gene1B"],
             help="Space-separated column names from the Fusions TSV where to update gene names \
```
Comment (dmiller15) [inline scripts/update_gene_symbols.py]:
```suggestion
            help="Space-separated column names from the Input TSV where to update gene names \
```

File: tools/update_gene_symbols.cwl [hunk 7]
```diff
@@ -32,11 +32,13 @@ inputs:
     inputBinding: { position: 1, prefix: "--new_symbol"} }
   update_columns: { type: 'string[]?', doc: "Space-separated column names from the Fusions TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults",
```
Comment (dmiller15) [inline tools/update_gene_symbols.cwl]:
```suggestion
  update_columns: { type: 'string[]?', doc: "Space-separated column names from the Input TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults",
```

File: workflows/dgd_fusion_annotator_wf.cwl [hunk 8]
```diff
@@ -10,7 +10,7 @@ inputs:
   output_basename: { type: string, doc: "Basename for the output TSV file" }
   # update gene symbols
   hgnc_tsv: { type: 'File?', doc: "Gene name database TSV file from HGNC. i.e. hgnc_complete_set.txt. If not provided, updated will be skipped" }
-  fusions_tsv: { type: File, doc: "Custom Fusions TSV file, i.e. fusion-dgd.tsv.gz" }
+  input_tsv: { type: File, doc: "Input TSV file, i.e. fusion-dgd.tsv.gz" }
   old_symbol: { type: 'string?', doc: "Column name for the old gene symbol(s) in the HGNC TSV. Set to override script defaults" }
   new_symbol: { type: 'string?', doc: "Column name for the new gene symbol(s) in the HGNC TSV. Set to override script defaults" }
   update_columns: {type: 'string[]?', doc: "Column names from the Fusions TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults" }
```
Comment (dmiller15) [inline workflows/dgd_fusion_annotator_wf.cwl]:
```suggestion
  update_columns: {type: 'string[]?', doc: "Column names from the Input TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults" }
```

**Review by dmiller15 — CHANGES_REQUESTED** [review]
You missed a lot of mentions of "fusion" in the update_fusion_gene_symbols.py script. I've pointed out a few here but there's even more.

File: scripts/update_gene_symbols.py (hunk 2 above)
Comment (migbro) [reply scripts/update_gene_symbols.py:142]: No, I don't actually want that. If the input did not have headers to begin with, then the output should remain as such. This is simply to allow the downstream functions to know which columns to act on

**Review by jharenza — APPROVED** [review]

**Review by dmiller15 — APPROVED** [review]
👍 Looks great!

**Pull request merged**